extensions/authentication-passport/src/strategy-adapter.ts

```diff
@@ -42,9 +42,9 @@ export class StrategyAdapter<U> implements AuthenticationStrategy {
    *     3. authenticate using the strategy
    * @param request The incoming request.
    */
-  authenticate(request: Request): Promise<UserProfile> {
+  authenticate(request: Request): Promise<UserProfile | RedirectRoute> {
```

Comment:
@deepakrkris could you explain more about why we have to allow returning a `RedirectRoute` here?

According to the [passport-oauth2](https://github.com/jaredhanson/passport-oauth2#configure-strategy) module, it returns a user (which conceptually same as lb's `userProfile`) after the authentication done, wondering why we couldn't follow the same pattern here?

Reply:
The `configure-strategy` deep-link you have commented above, is only to set the list of strategies in the base 'passport' module (which is heavily dependent on express).
The `passport.authenticate()` method that is handling the [express routes] (https://github.com/jaredhanson/passport-oauth2#authenticate-requests) interprets the incoming request and does the redirection. The base 'passport' module implements this as an express middleware without any code written by the user. For LoopBack, the strategy adapter will have to handle the redirect callback from the passport-strategy (done by the base `passport` module in express)  + we also need the controller to configure the routes (done by the `passport.authenticate()` method) + the injected send() method in the sequence must be aware of this redirection (done by the base passport module in express).

Reply:
This is the part I found tricky (due to my lack of knowledge in authentication when I started in Jan 2019), the authentication strategy code needs to execute before it reaches a controller method. If the user is not allowed to call this controller method, then they should get a `not authenticated` http error.  So in the flow you're following in all the code updates you've made @deepakrkris , is a user able to access and get into controller method code `before` this oauth authentication flow is finished? Perhaps we need a chart explaining each step of the code path through our code and through the oauth provider backend?

Reply:
@jannyHou @emonddr users are able to access the oauth2controller methods before authentication completes because they participate to get the user validated with an external system. Those methods are essentially for authenticating the user and are small parts of the larger oauth2 dialog.

Reply:
the Oauth2Controller methods that we have defined are similar to the endpoints defined in the [authenticate requests](https://github.com/jaredhanson/passport-oauth2#authenticate-requests) section of the passport oauth2 readme.

Reply:
@deepakrkris Thank you for the explanation 👍 the redirection passes the authenticated user's info to the callback endpoint, which results in the controller function maps to `/thirdparty`  won't have a user profile.